codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/transformers/RefactorConstrainedMemberType.kt

```diff
@@ -0,0 +1,195 @@
+import software.amazon.smithy.model.traits.Trait
+import software.amazon.smithy.model.transform.ModelTransformer
+import software.amazon.smithy.rust.codegen.core.smithy.DirectedWalker
+import software.amazon.smithy.rust.codegen.core.util.orNull
+import software.amazon.smithy.rust.codegen.server.smithy.allConstraintTraits
+import software.amazon.smithy.rust.codegen.server.smithy.hasConstraintTrait
+import software.amazon.smithy.rust.codegen.server.smithy.traits.RefactoredStructureTrait
+import software.amazon.smithy.utils.ToSmithyBuilder
+import java.lang.IllegalStateException
+import java.util.*
+
+object RefactorConstrainedMemberType {
+    private data class MemberTransformation(
+        val newShape: Shape,
+        val memberToChange: MemberShape,
+        val traitsToKeep: List<Trait>,
+    )
+
+    /**
+     * Transforms all member shapes that have constraints on them into equivalent
+     * non-constrained member shapes.
+     */
+    fun transform(model: Model): Model {
+        val existingRefactoredNames = HashSet<ShapeId>()
+
+        val transformations = model.operationShapes
+            .flatMap { operation ->
+                listOfNotNull(operation.input.orNull(), operation.output.orNull())
+            }
+            .map { model.expectShape(it) }
+            .flatMap {
+                val walker = DirectedWalker(model)
+                walker.walkShapes(it)
+            }
+            .filterIsInstance(StructureShape::class.java)
+            .flatMap {
+                it.constrainedMembers()
+            }.mapNotNull { it.makeNonConstrained(model, existingRefactoredNames) }
+
+        return applyTransformations(model, transformations)
+    }
+
+    /***
+     * Returns a Model that has all the transformations applied on the original
+     * model.
+     */
+    private fun applyTransformations(
+        model: Model,
+        transformations: List<MemberTransformation>,
+    ): Model {
+        if (transformations.isEmpty())
+            return model
+
+        val modelBuilder = model.toBuilder()
+        val memberShapesToChange: MutableList<MemberShape> = mutableListOf()
+
+        transformations.forEach {
+            modelBuilder.addShape(it.newShape)
+
+            val changedMember = it.memberToChange.toBuilder()
+                .target(it.newShape.id)
+                .traits(it.traitsToKeep)
+                .build()
+            memberShapesToChange.add(changedMember)
+        }
+
+        // Change all original constrained member shapes with the new standalone types,
+        // and keep only the non-constraint traits on the member shape.
+        return ModelTransformer.create()
+            .replaceShapes(modelBuilder.build(), memberShapesToChange)
+    }
+
+    /**
+     * Returns a list of members that have constraint traits applied to them
+     */
+    private fun StructureShape.constrainedMembers(): List<MemberShape> =
+        this.allMembers.values.filter {
+            it.hasConstraintTrait()
```

Comment:
This is transforming member shapes which have `@required` on them, which we don't want to touch (see the definition of `hasConstraintTrait`).

Add a negative test so that we don't transform these.